winch(x64): Add support for `loop`, `br` and `br_if`

saulecabrera opened the pull request:
Part of https://github.com/bytecodealliance/wasmtime/issues/6528

This change adds support for the `loop`, `br` and `br_if` instructions as well as unreachable code handling. Whenever an instruction that affects reachability is emitted (`br` in the case of this PR), the compiler will enter into an unreachable code state, essentially ignoring most of the subsequent instructions. When handling the unreachable code state some instructions are still observed, in order to determine if reachability should be restored.

This change, particulary the handling of unreachable code, adds all the necessary building blocks to the compiler to emit other instructions that affect reachability (e.g `unreachable`, `return`).

I decided to bundle these instructions together to be able to test more complex scenarios with `loop`, but I'm happy to split this change into two (or more). 

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

alexcrichton commented on winch/codegen/src/codegen/mod.rs:
```diff
@@ -101,7 +182,13 @@ where
 
         while !body.eof() {
             let offset = body.original_position();
-            body.visit_operator(&mut ValidateThenVisit(validator.visitor(offset), self))??;
+            if self.context.reachable {
+                body.visit_operator(&mut ValidateThenVisit(validator.visitor(offset), self))??;
+            } else {
+                let op = body.read_operator()?;
+                validator.op(offset, &op)?;
+                self.handle_unreachable_state(&op);
+            }
```
Bit of a drive-by comment, but altering the core loop here felt a bit unfortunate mainly along the lines of the translation of `If` for example was spread across `handle_unreachable_state` as well as the typical visitor.

One option though might be to use macro-magic to get `ValidateThenVisit` to skip all instructions except the allow-listed control flow instructions when the state is `unreachable`, sort of like how the visitor today fails on all instructions except for an allow-listed set that's implemented?

saulecabrera replied:
If the macro approach doesn't work and if there's a strong preference to avoid mixing the visitors with match-on-op, @cfallin's suggestion is what I had in mind as a fallback. One thing that I like of the current approach, is that the handling of unreachable code is isolated and should only be updated if there's a need to deal with such state while handling the unreachable code state in each visitor function is a bit more manual and error prone, but in the spirit of more explicit code I think that's fine.

alexcrichton replied:
Also to be clear I don't feel strongly about this at all, so whatever feels like it works best I think is fine

cfallin commented on winch/codegen/src/codegen/control.rs:
```diff
@@ -92,9 +144,30 @@ impl ControlStackFrame {
-            original_stack_size: 0,
+            original_stack_len: 0,
             result,
+            branch_target: false,
```
Can we call this field something more like `exit_reachable`? `branch_target` makes me think this is a label; really this is a reachability flag for some other program point than the current one.

cfallin replied:
Actually, after reading further: I was struggling to understand how this got set when the block "falls through"; it turns out it it isn't, and rather that case is implicitly handled because if we reach an `End` for a block while in reachable code, we'll remain reachable. So maybe something like `is_branch_target` makes it clear it's a predicate.

saulecabrera replied:
Updated!

cfallin commented on winch/codegen/src/codegen/mod.rs:
```diff
@@ -179,12 +266,24 @@ where
+    pub fn control_at(frames: &mut [ControlStackFrame], depth: u32) -> &mut ControlStackFrame {
+        let index = (frames.len() - 1)
+            .checked_sub(depth as usize)
+            .unwrap_or_else(|| panic!("expected valid control stack frame at index: {}", depth));
```
Does this need to handle `implicit` frames? Or is there an invariant here that only the outer body frame is implicit? If so, maybe we could remove the `implicit` flag (since it allows representing state that violates invariants) and instead handle the body-block case another way (e.g. at end/pop, recognize that it's the outermost block)?

saulecabrera replied:
Yeah, really the only implicit frame that should exist should be the outermost. So this flag can be removed I believe and instead use the "outermost" frame as the invariant here.

saulecabrera replied:
Updated!

cfallin reviewed (commented):
A few thoughts but overall I like how this is turning out -- the abstractions around the `ControlStackFrame` are very clean!

saulecabrera commented:
@cfallin I've addressed all of your comments, I think this is ready for another pass. Regarding the visitor vs match-on-op discussion above, I've updated the approach to rely only on the visitor approach, by extending the existing `validate_then_visit` macro as suggested by @alexcrichton.

saulecabrera commented on winch/codegen/src/codegen/mod.rs:
```diff
@@ -77,9 +77,90 @@ where
+    /// code state. Most of the instructions are dropped, except for certain
+    /// control instructions, which are observed in order to ensure that
+    /// reachability is correctly restored.
+    fn handle_unreachable_state(&mut self, op: &Operator) {
```
Updated to a new name according to the new approach for handling unreachable operators.

alexcrichton commented on winch/codegen/src/codegen/mod.rs:
```diff
@@ -77,9 +77,82 @@ where
+    ///
+    /// When the compiler is in an unreachable state, all the other instructions are not
+    /// visited.
+    pub fn handle_unreachable_else_or_end_operator(&mut self, op: &Operator) {
```
If this is only ever used for `End` and `Else` and never anything else, then it may be worth splitting this into two functions?

cfallin replied:
Agreed, let's split these cases out, like the other helpers.

alexcrichton reviewed (commented):
Nice! Seems good to me but I'd defer to @cfallin still for other technical review bits

cfallin commented on winch/codegen/src/codegen/mod.rs:
```diff
@@ -113,16 +186,38 @@ where
                         self.0.$visit($($($arg.clone()),*)?)?;
-                        Ok(self.1.$visit($($($arg),*)?))
+                        // Only visit operators if the compiler is in a reachable code state. If
+                        // the compiler is in an unrechable code state, most of the operators are
```
s/unrechable/unreachable/

cfallin commented on winch/codegen/src/codegen/mod.rs:
```diff
@@ -113,16 +186,38 @@ where
+                        // ignored except for If, Block, Loop, Else and End. These operators need
+                        // to be observed in order to keep the control stack frames balanced and to
+                        // determine if reachability should be restored.
+                        if self.1.is_reachable() || matches!(stringify!($op), "If" | "Block" | "Loop" | "Else" | "End") {
```
Will this do a string-compare at runtime? I think it would be better to have a predicate on the opcode (actual enum value) that returns a bool -- call it something like `visit_op_when_unreachable` -- and use it here.

saulecabrera replied:
Good call, updated!

cfallin reviewed (approved):
Generally LGTM, thanks! A few comments below but feel free to merge when addressed.

saulecabrera commented:
Seems that the Fuzz Targets step is having issues trying to fetch OCaml:

```
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/o/ocaml/ocaml-compiler-libs_4.13.1-3ubuntu1_amd64.deb  Error reading from server. Remote end closed connection [IP: 52.154.174.208 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/o/ocaml/ocaml-interp_4.13.1-3ubuntu1_amd64.deb  Error reading from server. Remote end closed connection [IP: 52.154.174.208 80]
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/universe/o/ocaml/ocaml_4.13.1-3ubuntu1_amd64.deb  Error reading from server. Remote end closed connection [IP: 52.154.174.208 80]
```

saulecabrera commented:
Seems to be fixed!
